test/integration/listener_lds_integration_test.cc

```diff
@@ -409,5 +409,66 @@ TEST_P(ListenerIntegrationTest, MultipleLdsUpdatesSharingListenSocketFactory) {
   }
 }
 
+TEST_P(ListenerIntegrationTest, ChangeListenerAddress) {
+  on_server_init_function_ = [&]() {
+    createLdsStream();
+    sendLdsResponse({MessageUtil::getYamlStringFromMessage(listener_config_)}, "1");
+    createRdsStream(route_table_name_);
+  };
+  initialize();
+  test_server_->waitForCounterGe("listener_manager.lds.update_success", 1);
+  // testing-listener-0 is not initialized as we haven't pushed any RDS yet.
+  EXPECT_EQ(test_server_->server().initManager().state(), Init::Manager::State::Initializing);
+  // Workers not started, the LDS added listener 0 is in active_listeners_ list.
+  EXPECT_EQ(test_server_->server().listenerManager().listeners().size(), 1);
+  registerTestServerPorts({listener_name_});
+
+  const std::string route_config_tmpl = R"EOF(
+      name: {}
+      virtual_hosts:
+      - name: integration
+        domains: ["*"]
+        routes:
+        - match: {{ prefix: "/" }}
+          route: {{ cluster: {} }}
+)EOF";
+  sendRdsResponse(fmt::format(route_config_tmpl, route_table_name_, "cluster_0"), "1");
+  test_server_->waitForCounterGe(
+      fmt::format("http.config_test.rds.{}.update_success", route_table_name_), 1);
+  // Now testing-listener-0 finishes initialization, Server initManager will be ready.
+  EXPECT_EQ(test_server_->server().initManager().state(), Init::Manager::State::Initialized);
+
+  test_server_->waitUntilListenersReady();
+  // NOTE: The line above doesn't tell you if listener is up and listening.
+  test_server_->waitForCounterGe("listener_manager.listener_create_success", 1);
+  // Make a connection to the listener from version 1.
+  codec_client_ = makeHttpConnection(lookupPort(listener_name_));
+
+  // Change the listener address from loopback to wildcard.
+  const std::string new_address = ipVersion() == Network::Address::IpVersion::v4 ? "0.0.0.0" : "::";
+  listener_config_.mutable_address()->mutable_socket_address()->set_address(new_address);
+  sendLdsResponse({MessageUtil::getYamlStringFromMessage(listener_config_)}, "2");
+  sendRdsResponse(fmt::format(route_config_tmpl, route_table_name_, "cluster_0"), "2");
```

Comment:
Try to connect and verify that it fails before sendRdsResposne, for a warming listener after address change?

Reply:
I'm not sure if this is possible; from testing, the new port can't be registered until after `listener_manager.listener_create_success` is incremented. From my understanding, at this point the listener is already warmed?

Reply:
See e.g. [`onListenerWarmed`](https://github.com/envoyproxy/envoy/blob/v1.18.2/source/server/listener_manager_impl.cc#L737) - the stat is incremented within `addListenerToWorker`